[EuiPageTemplate] Ensure context provides updates to template children

mgadewoll opened the pull request:
## Summary

> ⚠️ This PR should be rebased with this [PR](https://github.com/elastic/eui/pull/7648) to ensure storybook stories for EuiPageTemplate are available

This PR is a follow-up to an issue noticed [here](https://github.com/elastic/eui/pull/7648#discussion_r1557146196).

`EuiPageTemplate` did not properly pass down props to the page template subcomponents.
The reason was that changes on the context did not actually trigger updates on the children using that context.

To ensure this, this PR adds a provider wrapper for `EuiPageTemplate` which ensures that 
a) all nested children AND the page template itself correctly consume the context and that 
b) there is a state update which triggers context updates inside the consumers.

## QA

- [ ] ensure that the staging [docs](https://eui.elastic.co/pr_7663/#/templates/page-template/examples) for `EuiPageTemplate` is the same as [production](https://eui.elastic.co/#/templates/page-template/examples)
- [ ] ensure that the staging storybook for `EuiPageTemplate` is the same as production (⚠️ waiting for previous [PR](https://github.com/elastic/eui/pull/7648) to be merged for this to be available)

### General checklist

- Browser QA
    - [ ] Checked in both **light and dark** modes
    - [ ] Checked in **mobile**
    - [ ] Checked in **Chrome**, **Safari**, **Edge**, and **Firefox**
    - [ ] Checked for **accessibility** including keyboard-only and screenreader modes
- Docs site QA
    - [ ] ~Added **[documentation](https://github.com/elastic/eui/blob/main/wiki/contributing-to-eui/documenting)**~
    - [x] Props have proper **autodocs** (using `@default` if default values are missing) and **[playground toggles](https://github.com/elastic/eui/blob/main/wiki/contributing-to-eui/documenting/playgrounds.md)**
    - [x] Checked **[Code Sandbox](https://codesandbox.io/)** works for any docs examples
- Code quality checklist - N/A
- Release checklist
    - [x] A **[changelog](https://github.com/elastic/eui/blob/main/wiki/contributing-to-eui/documenting/changelogs.md)** entry exists and is marked appropriately.
    - [ ] ~If applicable, added the **breaking change** issue label (and filled out the breaking change checklist)~

kibanamachine commented:
Preview staging links for this PR:
- Docs site: https://eui.elastic.co/pr_7663/
- Storybook: https://eui.elastic.co/pr_7663/storybook

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/1704)
* Commit: c2fbd231fcc75b668b43ea1197a37c3f447168ff

<!--buildkite-pr-comment-eui-pull-request-test-and-deploy
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/eui-pull-request-test-and-deploy/builds/1704","number":1704,"commit":"c2fbd231fcc75b668b43ea1197a37c3f447168ff"}],"number":1704}
buildkite-pr-comment-->

mgadewoll commented:
We addressed the updates in another PR and changed the approach to prevent having to use stateful wrapper. ([link](https://github.com/elastic/eui/pull/7648#pullrequestreview-1992635877))
